### ISSUE TITLE
Topological Sort says "There is likely a cycle", but I can't find it…

### ISSUE THREAD

**Issue by markandrus** [tophat/monodeploy#442]
**Describe the bug**

I have recently on-boarded a monorepo of shared code to monodeploy. I'm ecstatic! This should work better for me than lerna (for me) and let me use, e.g., the `wokspace:` protocol.

I have one issue, though: in my monorepo, which includes dev and non-dev dependencies between workspaces, monodeploy is claiming there is a cycle. I've manually traced my dependencies, and I have to say, I cannot find the cycle&hellip; I made a little [graphviz visualization](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%0A%20%20parsers%20%5Blabel%3D%22parsers%22%5D%0A%20%20authContext%20%5Blabel%3D%22auth-context%22%5D%0A%20%20comparisons%20%5Blabel%3D%22comparisons%22%5D%0A%20%20config%20%5Blabel%3D%22config%22%5D%0A%20%20cdkCommon%20%5Blabel%3D%22cdk-common%22%5D%0A%20%20environments%20%5Blabel%3D%22environments%22%5D%0A%20%20requestSigning%20%5Blabel%3D%22request-signing%22%5D%0A%20%20timer%20%5Blabel%3D%22timer%22%5D%0A%20%20dynamo%20%5Blabel%3D%22dynamo%22%5D%0A%20%20release%20%5Blabel%3D%22release%22%5D%0A%20%20identifiers%20%5Blabel%3D%22identifiers%22%5D%0A%20%20o11y%20%5Blabel%3D%22o11y%22%5D%0A%20%20lambda%20%5Blabel%3D%22lambda%22%5D%0A%20%20testRunner%20%5Blabel%3D%22test-runner%22%5D%0A%20%20testUsers%20%5Blabel%3D%22test-users%22%5D%0A%20%20logging%20%5Blabel%3D%22logging%22%5D%0A%20%20%0A%20%20parsers%20-%3E%20identifiers%0A%20%20parsers%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20authContext%20-%3E%20identifiers%0A%20%20authContext%20-%3E%20parsers%0A%20%20authContext%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20comparisons%20-%3E%20identifiers%0A%20%20comparisons%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20config%20-%3E%20logging%0A%20%20config%20-%3E%20o11y%0A%20%20config%20-%3E%20timer%0A%20%20config%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20cdkCommon%20-%3E%20environments%0A%20%20%0A%20%20environments%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20requestSigning%20-%3E%20parsers%0A%20%20requestSigning%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20dynamo%20-%3E%20parsers%0A%20%20dynamo%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20identifiers%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20o11y%20-%3E%20identifiers%0A%20%20o11y%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20lambda%20-%3E%20authContext%0A%20%20lambda%20-%3E%20config%0A%20%20lambda%20-%3E%20logging%0A%20%20lambda%20-%3E%20o11y%0A%20%20lambda%20-%3E%20timer%0A%20%20lambda%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20%0A%20%20testUsers%20-%3E%20config%0A%20%20testUsers%20-%3E%20identifiers%0A%20%20testUsers%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%20%20testUsers%20-%3E%20identifiers%20%5Blabel%3Dtransitive%5D%0A%20%20testUsers%20-%3E%20parsers%20%5Blabel%3Dtransitive%5D%0A%20%20%0A%20%20logging%20-%3E%20identifiers%0A%20%20logging%20-%3E%20testRunner%20%5Blabel%3Ddev%5D%0A%7D) to confirm (cannot share the code, since it's private).

**To reproduce**

Sorry, I don't have a minimal repro yet, but I think the `visitedCount > maxPossibleVisits` check may be incorrect in some way.

**Expected behavior**

A monorepo with packages like those I've shared above releases successfully with the `--topological-dev` flag.

**Screenshots**

N/A

**Environment (please complete the following information):**

- OS: macOS
- Monodeploy Version: 2.8.13
- Node Version: 16

**Additional context**

```
➤ YN0001: │ Error: Unable to determine topological sort. There is likely a cycle.
    at getTopologicalSort (/home/circleci/project/node_modules/@monodeploy/dependencies/lib/getTopologicalSort.js:22:19)
    at publishPackages (/home/circleci/project/node_modules/@monodeploy/publish/lib/index.js:97:68)
    at /home/circleci/project/node_modules/@monodeploy/node/lib/index.js:121:53
    at StreamReport.startTimerPromise (/home/circleci/project/node_modules/@yarnpkg/core/lib/StreamReport.js:220:26)
    at pipeline (/home/circleci/project/node_modules/@monodeploy/node/lib/index.js:119:26)
    at async Function.start (/home/circleci/project/node_modules/@yarnpkg/core/lib/StreamReport.js:133:13)
    at async monodeploy (/home/circleci/project/node_modules/@monodeploy/node/lib/index.js:193:20)
    at async /home/circleci/project/node_modules/monodeploy/lib/cli.js:192:9
```

**Comment by noahnu** [comment]
I modelled your graph as a test and can reproduce:

```ts
    it('handles complex graphs', async () =>
        withMonorepoContext(
            {
                lambda: {
                    devDependencies: ['test-runner'],
                    dependencies: ['timer', 'config', 'o11y', 'auth-context', 'logging'],
                },
                timer: {},
                o11y: { devDependencies: ['test-runner'], dependencies: ['identifiers'] },
                config: { dependencies: ['timer', 'o11y', 'logging'] },
                logging: { devDependencies: ['test-runner'], dependencies: ['identifiers'] },
                'auth-context': {
                    devDependencies: ['test-runner'],
                    dependencies: ['identifiers', 'parsers'],
                },
                comparisons: { devDependencies: ['test-runner'], dependencies: ['identifiers'] },
                identifiers: { devDependencies: ['test-runner'] },
                'test-runner': {},
                parsers: { devDependencies: ['test-runner'], dependencies: ['identifiers'] },
                'request-signing': { devDependencies: ['test-runner'], dependencies: ['parsers'] },
                'test-users': {
                    devDependencies: ['test-runner'],
                    dependencies: ['config', 'identifiers'],
                },
                release: {},
                dynamo: { devDependencies: ['test-runner'], dependencies: ['parsers'] },
                'cdk-common': { dependencies: ['environments'] },
                environments: { devDependencies: ['test-runner'] },
            },
            async (context) => {
                const workspaces = [
                    'lambda',
                    'timer',
                    'o11y',
                    'config',
                    'logging',
                    'auth-context',
                    'comparisons',
                    'identifiers',
                    'test-runner',
                    'parsers',
                    'request-signing',
                    'test-users',
                    'release',
                    'dynamo',
                    'cdk-common',
                    'environments',
                ].map((k) => identToWorkspace(context, k))

                expect(await getTopologicalSort([...workspaces], { dev: true })).toEqual([])
            },
        ))
```

and a much smaller example:

```ts
    it('handles complex graphs', async () =>
        withMonorepoContext(
            {
                'pkg-1': {
                    dependencies: ['pkg-2', 'pkg-4'],
                },
                'pkg-2': {
                    dependencies: ['pkg-3', 'pkg-4'],
                },
                'pkg-3': { dependencies: ['pkg-4'] },
                'pkg-4': {},
            },
            async (context) => {
                const workspaces = ['pkg-1', 'pkg-2', 'pkg-3', 'pkg-4'].map((k) =>
                    identToWorkspace(context, k),
                )

                expect(await getTopologicalSort([...workspaces], { dev: true })).toEqual([
                    [workspaces[3] /* pkg-4 */],
                    [workspaces[2] /* pkg-3 */],
                    [workspaces[1] /* pkg-2 */],
                    [workspaces[0] /* pkg-1 */],
                ])
            },
        ))
```

If I remove the "cycle detection", it takes 87 iterations of the loop to produce a result.

So the cycle detection clearly doesn't work. Assuming the underlying implementation of the topological sort works, we need a new way to detect cycles (efficiently). I'm open to ideas/PRs. I'll have to think about this some more. Might need to loop back to this on Thursday though.

**Comment by markandrus** [comment]
@noahnu thanks for taking a look and repro-ing. I commented out the cycle check for now, and I've been able to adopt monodeploy into more of my projects.

**Comment by noahnu** [comment]
Fixed in monodeploy@2.8.14, however still open to improvements (including more test coverage) in the general algorithm.
